File: src/core/tokenizer.cpp
#include "core/tokenizer.hpp"

#include <cctype>

namespace search {

// ── Default English stop words ──────────────────────────────────────────────
// Kept intentionally lean: common function words that carry almost no
// discriminative value in an information-retrieval context.
Tokenizer::Tokenizer()
    : stop_words_{
          "a",      "an",     "the",    "is",     "are",    "was",
          "were",   "be",     "been",   "being",  "have",   "has",
          "had",    "do",     "does",   "did",    "will",   "would",
          "shall",  "should", "may",    "might",  "must",   "can",
          "could",  "of",     "in",     "to",     "for",    "on",
          "with",   "at",     "by",     "from",   "as",     "into",
          "through","during", "before", "after",  "above",  "below",
          "between","out",    "off",    "over",   "under",  "again",
          "further","then",   "once",   "and",    "but",    "or",
          "if",     "it",     "its",    "this",   "that",   "these",
          "those",  "i",      "me",     "my",     "we",     "our",
          "you",    "your",   "he",     "him",    "his",    "she",
          "her",    "they",   "them",   "their",  "what",   "which",
          "who",    "whom",   "not",    "no",     "nor",    "so",
          "very",   "too",    "here",   "there",  "when",   "where",
          "why",    "how",    "all",    "each",   "every",  "both",
          "few",    "more",   "most",   "other",  "some",   "such",
          "own",    "same",   "than",   "just",   "about"} {}

std::vector<std::string> Tokenizer::tokenize(std::string_view text) const {
    std::vector<std::string> tokens;
    std::string current;

    for (char ch : text) {
        if (std::isalnum(static_cast<unsigned char>(ch))) {
            // Accumulate alphanumeric characters, lowercasing on the fly.
            current += static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
        } else if (!current.empty()) {
            // Non-alphanumeric boundary → flush the current token.
            if (!remove_stop_words_ || !is_stop_word(current)) {
                tokens.push_back(std::move(current));
            }
            current.clear();
        }
    }

    // Flush any trailing token.
    if (!current.empty()) {
        if (!remove_stop_words_ || !is_stop_word(current)) {
            tokens.push_back(std::move(current));
        }
    }

    return tokens;
}

bool Tokenizer::is_stop_word(const std::string& token) const {
    return stop_words_.contains(token);
}

} // namespace search

File: api/engine/tokenizer.py
"""
tokenizer.py — Text tokenizer with lowercasing, punctuation removal, and
               configurable stop-word filtering.

Ports: src/core/tokenizer.hpp / tokenizer.cpp

Pipeline:  split on non-alphanumeric → lowercase → drop stop-words.
           Mirrors C++ behaviour exactly: only [a-z0-9] tokens survive.
"""

from __future__ import annotations

import re

# ── Default English stop-word list ────────────────────────────────────────────
# Mirrors the hardcoded list in tokenizer.cpp.
_DEFAULT_STOP_WORDS: frozenset[str] = frozenset({
    "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
    "of", "with", "by", "from", "is", "was", "are", "were", "be", "been",
    "being", "have", "has", "had", "do", "does", "did", "will", "would",
    "could", "should", "may", "might", "shall", "can", "not", "no", "nor",
    "so", "yet", "both", "either", "neither", "each", "few", "more", "most",
    "other", "some", "such", "than", "then", "too", "very", "just", "that",
    "this", "these", "those", "it", "its", "we", "our", "you", "your",
    "he", "she", "they", "them", "their", "i", "me", "my", "us", "who",
    "what", "which", "when", "where", "why", "how", "all", "any", "if",
    "as", "up", "out", "about", "into", "through", "after", "before",
    "between", "during", "without", "within", "along", "following", "across",
    "behind", "beyond", "plus", "except", "until", "toward", "among",
})

# Compiled regex: replace any non-alphanumeric character with a space.
_NON_ALNUM = re.compile(r'[^a-z0-9]+')


class Tokenizer:
    """
    Stateless text tokenizer.

    Time:  O(N) where N = len(text)
    Space: O(T) where T = number of output tokens
    """

    def __init__(self) -> None:
        self._stop_words: frozenset[str] = _DEFAULT_STOP_WORDS
        self._remove_stop_words: bool = True

    # ── Core API ──────────────────────────────────────────────────────────────

    def tokenize(self, text: str) -> list[str]:
        """
        Tokenize raw text into a list of normalised terms.

        Steps:
          1. Lowercase
          2. Replace all non-alphanumeric chars with spaces
          3. Split on whitespace
          4. Drop stop-words (if enabled) and empty tokens
        """
        lowered = text.lower()
        cleaned = _NON_ALNUM.sub(' ', lowered)
        tokens = cleaned.split()
        if self._remove_stop_words:
            tokens = [t for t in tokens if t not in self._stop_words]
        return tokens

    # ── Configuration ─────────────────────────────────────────────────────────

    def enable_stop_word_removal(self, enable: bool) -> None:
        self._remove_stop_words = enable

    def set_stop_words(self, words: set[str]) -> None:
        self._stop_words = frozenset(words)
